src/bscPlugin/codeActions/ScopeBrsFileCodeActionsProcessor.ts

```diff
@@ -0,0 +1,57 @@
+import type { CodeAction, Range } from 'vscode-languageserver';
+import { CodeActionKind } from 'vscode-languageserver';
+import { codeActionUtil } from '../../CodeActionUtil';
+import type { DiagnosticMessageType } from '../../DiagnosticMessages';
+import { DiagnosticCodeMap } from '../../DiagnosticMessages';
+import type { BrsFile } from '../../files/BrsFile';
+import type { BsDiagnostic } from '../../interfaces';
+import type { Scope } from '../../Scope';
+import { util } from '../../util';
+
+export class ScopeBrsFileCodeActionsProcessor {
+    public constructor(
+        public scope: Scope,
+        public file: BrsFile,
+        public range: Range,
+        public diagnostics: BsDiagnostic[],
+        public codeActions: CodeAction[]
+    ) {
+
+    }
+
+    public process() {
+        const isBsFile = this.file.extension.endsWith('.bs');
+        for (const diagnostic of this.diagnostics) {
+            //certain brighterscript-specific code actions
+            if (isBsFile) {
+                if (diagnostic.code === DiagnosticCodeMap.callToUnknownFunction) {
+                    this.suggestImports(diagnostic as any);
+                }
+            }
+        }
+    }
+
+    private suggestImports(diagnostic: DiagnosticMessageType<'callToUnknownFunction'>) {
```

Comment:
So, the the only "issue" with this function is that it will generate a code action for every scope, which are all identical. I don't have a great way of preventing that on this side, so @elsassph I'm up for any suggestions you might have. 

I could handle this in the onProgramGetDiagnostics, and look for the FIRST occurance of this diagnostic on any scope, but that seems like excess diagnostic filtering and scope looping. but then again, this might be more expensive since there could be many scopes with this problem.

codeActions are also done on demand, so combined with the diagnostic deduping, perhaps it's not worth working around right now. it's still really fast...

Reply:
Ahem, reminds me about our discussion about whether there should be scope-based actions at all.

Reply:
Yep, you won me over. :) I refactored the code to only have a single program event now.